zjit/src/hir.rs

```diff
@@ -2463,53 +2466,65 @@ impl Function {
+                                        // No (monomorphic/skewed polymorphic) profile info
                                         self.push_insn_id(block, insn_id); continue;
+                                    };
+                                    self.push_insn(block, Insn::PatchPoint { invariant: Invariant::MethodRedefined { klass, method: mid, cme }, state });
+                                    if klass.instance_can_have_singleton_class() {
+                                        self.push_insn(block, Insn::PatchPoint { invariant: Invariant::NoSingletonClass { klass }, state });
                                     }
-                                }
-                                // Get the profiled type to check if the fields is embedded or heap allocated.
-                                let Some(is_embedded) = self.profiled_type_of_at(recv, frame_state.insn_idx).map(|t| t.flags().is_struct_embedded()) else {
-                                    // No (monomorphic/skewed polymorphic) profile info
+                                    if let Some(profiled_type) = profiled_type {
+                                        recv = self.push_insn(block, Insn::GuardType { val: recv, guard_type: Type::from_profiled_type(profiled_type), state });
+                                    }
+                                    // All structs from the same Struct class should have the same
+                                    // length. So if our recv is embedded all runtime
+                                    // structs of the same class should be as well, and the same is
+                                    // true of the converse.
+                                    //
+                                    // No need for a GuardShape.
+                                    let replacement = if is_embedded {
+                                        let offset = RUBY_OFFSET_RSTRUCT_AS_ARY + (SIZEOF_VALUE_I32 * index);
+                                        if let (OptimizedMethodType::StructAset, &[val]) = (opt_type, args.as_slice()) {
+                                            self.push_insn(block, Insn::GuardNotFrozen { val: recv, state });
+                                            self.push_insn(block, Insn::StoreField { recv, id: mid, offset, val });
+                                            val
+                                        } else { // StructAref
+                                            self.push_insn(block, Insn::LoadField { recv, id: mid, offset, return_type: types::BasicObject })
+                                        }
+                                    } else {
+                                        let as_heap = self.push_insn(block, Insn::LoadField { recv, id: ID!(_as_heap), offset: RUBY_OFFSET_RSTRUCT_AS_HEAP_PTR, return_type: types::CPtr });
+                                        let offset = SIZEOF_VALUE_I32 * index;
```

Comment:
This is a little wonky -- can we split out the is_embedded/as_heap so that we don't have two GuardNotFrozen+StoreField or LoadField? I think we can re-assign recv as with the guardtype above